feat(allocator/vec2): align the `retain` method with the standard implementation

Dunqing opened the pull request:
The documentation and the implementation are copied from https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#2095-209.

~~After this change, `drain_filter` is no longer needed, so remove it.~~

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9752?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9752** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9752?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9752?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9742** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9742?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9679** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9679?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9655** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9655?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9730** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9730?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9729** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9728** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9752).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/vec2/mod.rs:
```diff
@@ -1295,26 +1295,33 @@ impl<'bump, T: 'bump> Vec<'bump, T> {
-    /// let b = Bump::new();
-    ///
-    /// let mut vec = bumpalo::vec![in &b; 1, 2, 3, 4];
+    /// let mut vec = vec![1, 2, 3, 4];
```
`vec!` macro is for `std::vec::Vec`.

```suggestion
    /// use bumpalo::Bump;
    /// let bump = Bump::new();
    /// let mut vec = Vec::from_iter_in([1, 2, 3, 4], &bump);
```

Dunqing replied:
Good catch!

overlookmotel commented on crates/oxc_allocator/src/vec2/mod.rs:
````diff
@@ -1295,26 +1295,33 @@ impl<'bump, T: 'bump> Vec<'bump, T> {
+    /// external state may be used to decide which elements to keep.
+    ///
+    /// ```
+    /// let mut vec = vec![1, 2, 3, 4, 5];
````
Same as above.

overlookmotel reviewed (commented):
Do we want to remove `drain_filter`? It looks quite useful. e.g.:

```rs
// Move all statements except `import`s and `export`s into a nested `BlockStatement`
let stmts_to_move = program.body.drain_filter(|stmt| stmt.is_module_declaration());
let stmts_to_move = Vec::from_iter_in(stmts_to_move, ast.allocator);
program.body.push(ast.statement_block(SPAN, stmts_to_move));
```

This could make e.g. `using` transform more efficient:

https://github.com/oxc-project/oxc/blob/d13817e1040bd1b0159490d620ff3799945c81b3/crates/oxc_transformer/src/proposals/explicit_resource_management.rs#L218-L232

But if you want to remove it now, and we add it (or something like it) back again later, I'm fine with that.

Dunqing commented:
> Do we want to remove `drain_filter`? It looks quite useful. e.g.:
> 
> ```rust
> // Move all statements except `import`s and `export`s into a nested `BlockStatement`
> let stmts_to_move = program.body.drain_filter(|stmt| stmt.is_module_declaration());
> let stmts_to_move = Vec::from_iter_in(stmts_to_move, ast.allocator);
> program.body.push(ast.statement_block(SPAN, stmts_to_move));
> ```
> 
> This could make e.g. `using` transform more efficient:
> 
> https://github.com/oxc-project/oxc/blob/d13817e1040bd1b0159490d620ff3799945c81b3/crates/oxc_transformer/src/proposals/explicit_resource_management.rs#L218-L232
> 
> But if you want to remove it now, and we add it (or something like it) back again later, I'm fine with that.

Thank you for bringing this example up. I've checked the std library doesn't include this method, and doesn't remember where a case in our codebase can use this method, so I think it can be removed. Now considering there is a case that can benefit from this method, I will keep it.

overlookmotel commented:
### Merge activity

* **Mar 25, 10:05 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
feat(allocator/vec2): align the `retain` method with the standard implementation (#9752)

The documentation and the implementation are copied from https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#2095-209.

~~After this change, `drain_filter` is no longer needed, so remove it.~~
